feat(blog): using Tauri (#5203)
CloudFire-LLC/cloudfire: website/src/app/blog/using-tauri/readme.mdx

`connlib`) and TypeScript for the website and admin portal. I don't know much
web design, so using TypeScript allowed me to make a rough prototype of the GUI
and then let other team members polish and style it to match the website
branding.

JavaScript inside the web view can call into `async` Rust functions (TODO: Code
example), and from there we can use `tokio`'s mpsc channels to send events to a
central controller task. (TODO: Code)

# What snags did we hit?

Overall, I like Tauri a lot. This section is longer because there are lots of
little details. "It's good" is shorter than "We had trouble because..."

## The Tauri directory layout is odd

The root of a Tauri project, such as
[`/rust/gui-client` in the Firezone repo](https://github.com/firezone/firezone/tree/2a1187bd9c4d12960ef0000df1b6bd9bb361bf10/rust/gui-client),
has two subdirectories: `src` and `src-tauri`. The entire Rust project is
contained in `src-tauri`, and the `src` directory contains files for the web
view. `Cargo.toml` and `tauri.conf.json` are in `src-tauri`. I find this
confusing. I think Tauri's primary purpose is to wrap up web apps into native

> apps, so the Rust code is sequestered in `src-tauri`, and the project is

Reviewer: ```suggestion
apps, so the Rust code is isolated in `src-tauri`, and the project is
```

Maybe a bit simpler?

supposed to be an NPM project that incidentally contains a Rust wrapper.

> I think of the Firezone Client as a Rust GUI program that incidentally uses

Reviewer: ```suggestion
I think of the Firezone Client as a Rust program that incidentally uses
```

mentioning GUI on the next line might make this redundant.

HTML+JS+CSS for the GUI, so this layout is backwards to me. I would have
expected a top-level `/Cargo.toml`, with the Rust under `/src`, and the web
files under `/src-web` or similar. Maybe there's a reason for this, but I'm not
aware of it.

> ## Linux doesn't like running GUIs with root permissions

Reviewer: Interesting section! I think it's a good candidate for much more detail on the approaches we used for Windows and Linux specifically, and what other approaches we considered (I remember that design document you had drafted that had a few different options like single binary but run in different modes...)

and
[Windows](https://github.com/tauri-apps/tao/blob/f9f81f2fd761c60abb5b46a6469864562a811c6d/src/platform_impl/windows/event_loop.rs#L219).

I first suspected this is due to some platform-specific limitation, such as
Windows' `WinMain` function or the fact that some OSes require all GUI-related
function calls to happen on "the main thread", or at least on a single thread.

`tao` is a fork of `winit`, and `winit` is able to return from
[its run function](https://docs.rs/winit/latest/winit/event_loop/struct.EventLoop.html#method.run_app)
on desktop.

The
[`run_iteration` function](https://docs.rs/tauri/1.6.6/tauri/struct.App.html#method.run_iteration)
should be able to handle this, but instead it busy-loops.

When asked about these issues, lead Tauri developer Fabian said he also expects
better from the underlying OSes and this causes him to
["cry (himself) to sleep at night 🤷"](https://github.com/tauri-apps/tauri/issues/8631#issuecomment-1898328992)
Me too, Fabian, me too. /s

## Web views use a lot of RAM

This isn't Tauri's fault, it's a direct effect of using web technology.

The GUI process consistently uses about 100 MB of RAM even when idle. This seems

> to be the minimum for keeping web views open. Even if we hide all windows, it
> won't reduce the amount of RAM.

Reviewer: ```suggestion
to be the minimum for keeping the web view process alive. Even if we hide all windows, it
won't reduce the amount of RAM.
```

Was a bit confusing reading these two sentences together.

import Alert from "@/components/DocsAlert";
import AsciinemaPlayer from "@/components/AsciinemaPlayer";
import Image from "next/image";
import Link from "next/link";

> [I would have thought](https://github.com/tauri-apps/tauri/issues/8631#issuecomment-1896717039)
> after more than 20 years of GUI paradigms, the major OSes would have made GUIs
> as nice to program as CLIs. Oh well.

-- The author

> I thought so too at one point, until i got into working _on_ (instead of
> "with") a gui framework. Now I cry myself to sleep every night 🤷

> -- A lead Tauri developer

Reviewer: 😆

allows us to install and configure our background service by
[overriding the Wix files](https://github.com/firezone/firezone/blob/2a1187bd9c4d12960ef0000df1b6bd9bb361bf10/rust/gui-client/src-tauri/tauri.conf.json#L44-L45)
that generate the MSI.

`cargo-deb`
[does allow these scripts](https://github.com/kornelski/cargo-deb?tab=readme-ov-file#packagemetadatadeb-options),
but Tauri's bundler is based on
[`cargo-bundle`](https://github.com/burtonageo/cargo-bundle) instead, which
doesn't appear to support the scripts.

## The tray menu works, but it's odd

In Windows, you can't left-click on the tray menu. Tauri doesn't seem to fire
any event for it. This is not a limitation of Windows, so I'm not sure why Tauri
only allows right clicks.

On GNOME, if the menu is too long and the screen is too short, you can't open
submenus. If the menu is barely too long, the submenus open but they're squashed
into a tiny scroll view that doesn't look like a submenu:

<Image
  width={888}
  height={724}
  src="/images/blog/using-tauri/squashed-menu.png"
  alt="A very long menu displayed in GNOME, with the submenu being squashed"

> />

Reviewer: ```suggestion

<Link href="/images/blog/using-tauri/squashed-menu.png" target="_blank">
  <Image
    width={888}
    height={724}
    src="/images/blog/using-tauri/squashed-menu.png"
    alt="A very long menu displayed in GNOME, with the submenu being squashed"
  />
</Link>
```

I sometimes wrap these in Links so that folks can click them to expand. Up to you though.

On Windows, the submenus look like submenus:

<Image
  width={1048}
  height={654}
  src="/images/blog/using-tauri/good-menu.png"
  alt="A very long menu displayed in GNOME, with the submenu being squashed"

> />

Reviewer: ```suggestion
<Link href="/images/blog/using-tauri/good-menu.png" target="_blank">
  <Image
    width={1048}
    height={654}
    src="/images/blog/using-tauri/good-menu.png"
    alt="A very long menu displayed in GNOME, with the submenu being squashed"
  />
</Link>
```

| [native-windows-gui](https://crates.io/crates/native-windows-gui)         | Looked easy                                                | Not maintained                                                                                               |
| [WPF](https://en.wikipedia.org/wiki/Windows_Presentation_Foundation)      | Small binaries, I've used it before                        | Binding C# to Rust considered infeasible, assumed Rust and Linux support to be bad                           |
| [Electron](<https://en.wikipedia.org/wiki/Electron_(software_framework)>) | Same positives as Tauri, but even more popular             | Infamously large download sizes                                                                              |
| [Iced](https://github.com/iced-rs/iced)                                   | Smaller downloads and less RAM usage than Tauri / Electron | Hasn't caught on yet, had some trouble with compatibility when building in Ubuntu 20.04 and running in 22.04 |

> # What's so good about Tauri?

Reviewer: ```suggestion
## What's so good about Tauri?
```

> ## Tauri is easy to learn and use

Reviewer: ```suggestion
### Tauri is easy to learn and use
```

beginners into advanced users and contributors. Python became popular by
appealing to new programmers who needed to write small programs. JavaScript
became popular by being pre-installed in every web browser. C++ became popular
by being published before both of them.

Tauri makes itself easy by shipping the `create-tauri-app` tool for setting up
boilerplate projects, and by having bundlers for Windows `.msi` installers and
Debian `.deb` packages that Just Work, at least for simple programs.

<AsciinemaPlayer
  src="/asciinema/tauri.cast"
  autoplay={false}
  rows={30}
  idleTimeLimit={3}
  preload={true}
/>

Using Tauri, we got the Windows Client to beta in about 2 months, including the
time needed for me to learn about `connlib`, and for the team to change our auth
flow from one that worked well on macOS to one that worked on all platforms.
This has been my first project using Tauri, and the pace of development has been
great.

Bringing the Linux GUI Client to MVP took another 4 months. This took longer
than Windows because Linux's security model required us to split the Client into
two processes, and because we refactored the Windows Client to share the code
with Linux.

> ## We already use Rust and TypeScript

Reviewer: ```suggestion
### We already use Rust and TypeScript
```

Firezone already uses Rust for our data plane (the Gateway, the Relay, and
`connlib`) and TypeScript for the website and admin portal, so we didn't have to
add another language like C++ or C# to the mix. I made a prototype of the GUI in
TypeScript, hooked it up to the Rust code, and then let another team member
polish and style it to match the website branding.

<Link href="/images/blog/using-tauri/firezone-3a4f2cf.jpeg" target="_blank">
  <Image
    width={1339}

>     height={555}

Reviewer: ```suggestion
    height={555}
    className="rounded shadow"
```

    src="/images/blog/using-tauri/firezone-3a4f2cf.jpeg"
    alt="The Windows Client shortly before MVP, 2024 Jan 04"
  />
</Link>

_The Windows Client shortly before MVP, 2024 Jan 04_

<Link href="/images/blog/using-tauri/firezone-926e26f57.jpeg" target="_blank">
  <Image
    width={2694}

>     height={1090}

Reviewer: ```suggestion
    height={1090}
    className="rounded shadow"
```

    auth_base_url: authBaseUrlInput.value,
    api_url: apiUrlInput.value,
    log_filter: logFilterInput.value,
  },
});
```

[The Rust implementation (settings.rs#L47-L61)](https://github.com/firezone/firezone/blob/37b3bc2047f6918adecda4108708b9ba9f76b84b/rust/gui-client/src-tauri/src/client/settings.rs#L47-L61)

```rust
#[tauri::command]
pub(crate) async fn apply_advanced_settings(
    managed: tauri::State<'_, Managed>,
    settings: AdvancedSettings,
) -> Result<(), String> {
    if managed.inner().inject_faults {
        tokio::time::sleep(Duration::from_secs(2)).await;
    }
    apply_inner(&managed.ctlr_tx, settings)
        .await
        .map_err(|e| e.to_string())?;

    Ok(())
}
```

Tauri automatically serializes and deserializes both ways as long as our types
implement the common `serde::{Deserialize, Serialize}` traits, which we
`derive`:

[settings.rs#L12](https://github.com/firezone/firezone/blob/37b3bc2047f6918adecda4108708b9ba9f76b84b/rust/gui-client/src-tauri/src/client/settings.rs#L12)

```rust
#[derive(Clone, Deserialize, Serialize)]
pub(crate) struct AdvancedSettings {
    pub auth_base_url: Url,
    pub api_url: Url,
    pub log_filter: String,
}
```

> ## Tauri comes with batteries included

Reviewer: ```suggestion
### Tauri comes with batteries included
```

  browsers, like how Steam or Zoom can be triggered by links

Tauri has built-in modules for a
[system tray menu](https://docs.rs/tauri/1.6.7/tauri/struct.SystemTrayMenu.html),
[notifications](https://docs.rs/tauri/1.6.7/tauri/api/notification/struct.Notification.html),
and [self-updates](https://docs.rs/tauri/1.6.7/tauri/updater/index.html), and it
has a `.deb` [bundler for Linux](https://tauri.app/v1/api/config/#debconfig)
built in. For deep links we used
[tauri-plugin-deep-link](https://crates.io/crates/tauri-plugin-deep-link).

This made it easy to commit to Tauri - We knew that worst-case, we had
_something_ that would work for a beta release.

Some of those batteries stayed in and some got replaced. We replaced the default
Tauri notifications on Windows with `tauri-winrt-notification`, we rewrote the
deep link code, we haven't deployed self-updates yet, and the system tray menu
will be replaced eventually.

Over the last 6 months Tauri's built-in features have saved us:

- From using system tray and notification APIs directly
- From starting from scratch for deep linking
- From writing an updater at all

> # How does Tauri work for us?

Reviewer: ```suggestion
## How does Tauri work for us?
```

```text
Firezone Client
    |
    +------> Tokio
    |
    +------> connlib
    |
    +------> Tauri
```

The Firezone Client for Linux and Windows is a wrapper around our cross-platform
connection library `connlib`. Tauri provides the GUI and bundling, and then we
use [Tokio](https://github.com/tokio-rs/tokio), the popular async runtime for
Rust, to glue everything together.

> ## Tauri = TAO + WRY

Reviewer: ```suggestion
### Tauri = TAO + WRY
```

```text
UI process
    |
    +--------> WebContent process
    |
    +--------> WebContent process
    |
    +--------> WebContent process
```

(Paraphrased from
[webkit.org](https://docs.webkit.org/Getting%20Started/Introduction.html#what-is-webkit))

The two engines are essentially cousins as well:

- In 2001, [Apple forked KHTML](https://en.wikipedia.org/wiki/WebKit#Origins) to
  make WebKit, the browser engine for Safari.
- In 2008, Chromium 1.0 released, based on a
  [multi-process branch of WebKit](<https://en.wikipedia.org/wiki/Chromium_(web_browser)#Architecture>)
- In 2013, Chromium forked WebKit and named it
  [Blink](<https://en.wikipedia.org/wiki/Blink_(browser_engine)>)
- In 2015, Microsoft released
  [Edge](https://en.wikipedia.org/wiki/Microsoft_Edge) using an engine that no
  longer exists
- In 2020, Microsoft released the new Edge, built atop Chromium, and they
  released
  [WebView2](https://blogs.windows.com/msedgedev/2020/10/19/edge-webview2-general-availability/)

So WebView2 is an API for Microsoft Edge, which is Google Chromium, based on
Google Blink, which was WebKit, meant for Apple Safari, originally KDE KHTML.

Firezone is best viewed with

> [Konqueror](https://en.wikipedia.org/wiki/Konqueror).

Reviewer 2: :joy:

<Link href="/images/blog/using-tauri/konqueror.png" target="_blank">
  <Image
    width={1024}

>     height={768}

Reviewer: ```suggestion
    height={768}
    className="rounded shadow"
```

    src="/images/blog/using-tauri/konqueror.png"
    alt="The primordial soup from which the Web evolved"
  />
</Link>

_The primordial soup from which the Web evolved_

(Source:
https://en.wikipedia.org/wiki/K_Desktop_Environment_2#/media/File:KDE_2.2.2.png)

> ## Bundling 📦

Reviewer: ```suggestion
### Bundling 📦
```

webkit2gtk, so `apt` will install it for us when Firezone is installed, share it
with other software on the system, and upgrade it when security patches are
released.

```text
deb package
    |
    +----> firezone-client-gui
    |
    +----> systemd service
    |
    +----> HTML+JS+CSS assets
    |
    +----> dependency on webkit2gtk
```

For Windows, Tauri creates an `.msi` installer, which automatically downloads
and installs WebView2 if the system doesn't have it already. Windows 11 systems
have WebView2 pre-installed, and just like `apt` on Ubuntu, we let Windows
handle the security updates.

```text
MSI installer
    |
    +----> firezone-client-gui.exe
    |
    +----> Windows service
    |
    +----> HTML+JS+CSS assets
    |
    +----> WebView2 downloader
```

Firezone Client 1.0.5 is **12 MB on Linux** and **8 MB on Windows**, much
smaller than the default ~100 MB for Electron, and still a bit smaller than the
~35 MB achievable with
[electron-builder](https://github.com/electron-userland/electron-builder).
Another API could have made it smaller, but considering how little effort it
takes to use Tauri, the download size is certainly near the Pareto front.

> # If Tauri is so great, why did the Linux port take twice as long as Windows?

Reviewer: ```suggestion
## If Tauri is so great, why did the Linux port take twice as long as Windows?
```

Reviewer: Generally the convention we've been using is `<h1>` for page header then `<h2>`s all around, followed by `<h3>`s etc

The Linux port took twice as long because running a GUI with `sudo` in Linux
doesn't seem to work, and we need root permissions to control DNS. On Windows,
the early releases were just one process running with admin privilege. (Not
counting the WebView renderer processes)

```text
+---------------------------+
|    Tauri GUI process      |
|                           |
|   +-------------------+   |
|   |   Secure tunnel   |   |
|   +-------------------+   |
+---------------------------+
```

When we ported that to Linux, everything broke.

The tray icon
[didn't appear on some Ubuntu versions](https://github.com/firezone/firezone/issues/4053),
so we couldn't click "Sign in", and when it did appear, we
[couldn't open a browser anyway](https://github.com/firezone/firezone/issues/4103),
because the desktop is expecting us to be a normal user and the GUI was running
as `root`.

So we gave up on that model and split the program into 2 top-level processes:

1. A GUI running as the normal desktop user
2. A systemd service running as root, with lots of sandboxing.

```text
+-----------------------+                                             +-------------------+

> |   Tauri GUI process   | <------ Inter-process communication ------> |   Secure tunnel   |

Reviewer 2: Should `Secure Tunnel` perhaps mention `connlib` here, just to be unambiguous?

Author: Yeah it should, I'll add that to the tasklist

means that the code running as root can't accidentally
[read your email](https://xkcd.com/1200/),
[delete the entire filesystem](https://hackaday.com/2024/01/20/how-a-steam-bug-once-deleted-all-of-someones-user-data/),
or even
[change the time on the system clock](https://github.com/firezone/firezone/blob/37b3bc2047f6918adecda4108708b9ba9f76b84b/rust/gui-client/src-tauri/deb_files/firezone-client-ipc.service#L19).

We copied this architecture to Windows, where it coincidentally fixed a bug
where
[our DNS control didn't deactivate](https://github.com/firezone/firezone/issues/4899)
if the system powered off suddenly or Firezone was force-stopped.

The Windows sandboxing is not as strict as systemd's, but on both platforms this
change was also:

- A convenience win, because you can run Firezone without admin privilege, and
- A bug-fixing win because the service can clean up Firezone's DNS control if
  the GUI crashes or the computer suddenly reboots.

So we got everything we wanted, just not in the order we wanted it.

> # What non-Tauri issues did we have?

Reviewer: ```suggestion
## What non-Tauri issues did we have?
```

> ## Web views use a whole bunch of RAM

Reviewer: ```suggestion
### Web views use a whole bunch of RAM
```

<Link href="/images/blog/using-tauri/lots-of-ram.png" target="_blank">
  <Image
    width={1598}

>     height={952}

Reviewer: ```suggestion
    height={952}
    className="rounded shadow"
```

    src="/images/blog/using-tauri/lots-of-ram.png"
    alt="Windows Task Manager showing that the webviews need about 100 MB of RAM to do nothing"
  />
</Link>

The GUI uses about 100 MB of RAM even with all windows hidden. This seems to be
the minimum for keeping the web view process alive. It's okay for an MVP, but
it's still 5% of the 2 GB needed for a minimum Ubuntu desktop, just to run a GUI
that will be closed most of the time.

We can't fix this right now, but we have at least 3 long-term options:

1. Try closing the windows instead of just hiding them
2. Only use the main GUI process to run the tray menu, and keep the Tauri
   windows in a subprocess that we can stop and restart when needed
3. Switch away from Tauri entirely. Iced uses about half as much RAM, and GTK+
   also has some Windows support and some Rust support.

Hopefully some of that 100 MB is code pages we can share with Edge and other
WebView2 apps, but I suspect most of it is HTML and JS overhead.

> ## Ubuntu forwards and backwards compatibility is hard

Reviewer: ```suggestion
### Ubuntu forwards and backwards compatibility is hard
```

we're running from an AppImage, and AppImages built on 20.04 have rendering bugs
on 24.04 anyway.

Switching to another GUI framework might work, but since GTK+ is already
involved in this incompatibility issue, it's likely that GTK+ itself has some
compatibility problems we'd hit. We'd have to spend time playing with it before
committing to a port. We also tried building an example Iced app on Ubuntu
20.04, and it didn't render on 24.04 either. Maybe it's something related to the
recent Wayland cut-over.

There is a plan that _should_ work, and it doesn't have any big leaps of faith,
but it's still painful:

- Refactor the code so that we can build for both Tauri 1.x and Tauri 2.x.
  (Needed sooner or later anyway)
- Use Docker or something to build one `.deb` inside 20.04 and another inside
  22.04. (Or worst case, two entire VMs)
- Unpack the two packages, add a Bash script shim that checks which WebKit is
  installed and launches the correct binary, then re-pack them into a single
  `.deb`. (We already re-pack Tauri's `.deb`, so this should be possible)

Or maybe we can install 4.1 on 20.04 somehow. Whatever we do, it won't be

> simple. Worst case, we can give up and use Electron.

Reviewer 2: Could make a pun here for writing our own GUI framework which seems to be the fun thing to do these days considering what is happening at Zed & linebender for example :)

Author: ha lol I've always wanted to but I'd rather see Firezone succeed so I keep getting paid :P

Author: Honestly I had wondered if we can get away with a fixed-grid thing and just basically be a terminal application that is halfway like a GUI, but that'll make localization outside of Latin-1 scripts difficult, and users will notice it looks weird

Like I'm pretty sure 99% of the difficulty in GUIs is that fonts and text are just impossible to get right and expensive to even attempt. I'm ready for a bitmap font revival

(The other 99% is getting stuff like text editing and accessibility just right, which ofc Chromium does for free and writing from scratch is awful)

> ## WebView2 just doesn't install sometimes

Reviewer: ```suggestion
### WebView2 just doesn't install sometimes
```

This is an upstream issue with WebView2 for Windows. Sometimes the installer
times out in our GitHub CI runners and we don't know why.

<Link href="/images/blog/using-tauri/webview-install-hang.png" target="_blank">
  <Image
    width={1230}

>     height={464}

Reviewer: ```suggestion
    height={464}
    className="rounded shadow"
```

Reviewer: This will make it consistent with the codeblocks, asciinema view, and other images

    src="/images/blog/using-tauri/webview-install-hang.png"
    alt="The WebView2 installer just hanging in CI for some reason"
  />
</Link>

https://github.com/firezone/firezone/pull/4981

> ## GNOME

Reviewer: ```suggestion
### GNOME
```

On GNOME, if the menu is too long and the screen is too short, you can't open
submenus. If the menu is barely too long, the submenus open but they're squashed
into a tiny scroll view. And the submenus are shown in-line, so at no point does
it look like a submenu at all:

<Link href="/images/blog/using-tauri/squashed-menu.png" target="_blank">
  <Image
    width={888}

>     height={724}

Reviewer: ```suggestion
    width={600}
    height={500}
    className="rounded shadow mx-auto"
```

This one looked a bit pixelated on my screen, so maybe reduce size and center. NextJS will handle proportional resizes so no need to get exact dimensions right.

Author: Oh I thought `width` and `height` was the size of the image as stored on disk. That makes a lot more sense.

Author: Turns out it _was_ pixelated, because I took the screenshot on the Mac host after macOS had applied scaling to qemu. I replaced it with a screenshot from inside the Ubuntu guest. Parallels seems to be hiDPI-aware and Qemu / UTM isn't.

    src="/images/blog/using-tauri/squashed-menu.png"
    alt="A very long menu displayed in GNOME, with the submenu being squashed"
  />
</Link>

On Windows, it's fine. It's been fine since Windows 95, and it's still fine.

<Link href="/images/blog/using-tauri/good-menu.png" target="_blank">
  <Image

>     width={1048}
>     height={654}

Reviewer: ```suggestion
    width={700}
    height={600}
    className="rounded shadow mx-auto"
```

Slightly pixelated,

    src="/images/blog/using-tauri/good-menu.png"

>     alt="A very long menu displayed in GNOME, with the submenu being squashed"

Reviewer 2: Outdated alt text.

  />
</Link>

Also in GNOME, clicking on a notification runs another instance of the app.
Maybe this is intended to activate the app, but it doesn't seem to happen on
Windows. The second instance doesn't get any obvious environment variable or
argument to hint that it was launched from a notification, so we show a generic
"Firezone is already running" error instead of doing anything useful. If we
change that error to do nothing, then users will be confused if they try to run
Firezone and don't notice it's already running.

> # What Tauri-related issues did we have?

Reviewer: ```suggestion
## What Tauri-related issues did we have?
```

> ## The Tauri directory layout is odd

Reviewer: ```suggestion
### The Tauri directory layout is odd
```

The root of a Tauri project, such as
[`/rust/gui-client` in the Firezone repo](https://github.com/firezone/firezone/tree/2a1187bd9c4d12960ef0000df1b6bd9bb361bf10/rust/gui-client),
has two subdirectories: `src` and `src-tauri`. The entire Rust project is
contained in `src-tauri`, and the `src` directory contains files for the web
view. `Cargo.toml` and `tauri.conf.json` are in `src-tauri`. I find this
confusing. Maybe Tauri's intended use case is to wrap up web apps into native
apps, so the web content gets to be in `src` and the Rust code is isolated over
in `src-tauri`.

But we think of the Firezone Client as a Rust program that incidentally uses
HTML+JS+CSS for the GUI, so this layout is backwards for us. We were expecting a
top-level `/Cargo.toml`, with the Rust code under `/src`, and the web files
under `/src-web` or something. There's probably a good reason why Tauri does
this, but we aren't sure what that reason is.

> ## Tauri's deb bundler is not very customizable

Reviewer: ```suggestion
### Tauri's deb bundler is not very customizable
```

   |          |
   |          +----> /usr/share/applications/firezone-client-gui.desktop
   |          |
   |          +----> /usr/share/icons/...
   |
   +----> control.tar.gz
              |
              +----> control
              |
              +----> md5sums
              |
              +----> postinst     (... but we have to tear up the build process)
              |                   (to add postinst and prerm here              )
              +----> prerm
```

`cargo-deb`
[does support these scripts](https://github.com/kornelski/cargo-deb?tab=readme-ov-file#packagemetadatadeb-options),
but Tauri's bundler is based on
[`cargo-bundle`](https://github.com/burtonageo/cargo-bundle) instead.

I'm surprised these scripts are not in Tauri's use cases. On Windows, we
[override Tauri's Wix files](https://github.com/firezone/firezone/blob/2a1187bd9c4d12960ef0000df1b6bd9bb361bf10/rust/gui-client/src-tauri/tauri.conf.json#L44-L45)
to generate an MSI that installs and configures our connlib service. So in both
cases we had to hijack the build process a little. Linux required more code,
Windows required more "ChatGPT, tell me how Wix works, please."

> ## You can't left-click on the tray menu in Windows

Reviewer: ```suggestion
### You can't left-click on the tray menu in Windows
```

In Windows, you can't left-click on the tray menu. Tauri doesn't fire any event
for it, even though other Windows apps handle left clicks just fine. You have to
right-click on Tauri system trays in Windows.

> ## Tauri apps can't exit gracefully

Reviewer: ```suggestion
### Tauri apps can't exit gracefully
```

[Linux](https://github.com/tauri-apps/tao/blob/f9f81f2fd761c60abb5b46a6469864562a811c6d/src/platform_impl/linux/event_loop.rs#L913)
and
[Windows](https://github.com/tauri-apps/tao/blob/f9f81f2fd761c60abb5b46a6469864562a811c6d/src/platform_impl/windows/event_loop.rs#L219).

This might be due to some platform-specific limitation, such as Windows'
`WinMain` function or the fact that some OSes require all GUI-related function
calls to happen on "the main thread", or at least on a single thread.

Interestingly, `tao` is a fork of `winit`, and `winit` is able to return from
[its run function](https://docs.rs/winit/latest/winit/event_loop/struct.EventLoop.html#method.run_app)
on desktop.

The
[`run_iteration` function](https://docs.rs/tauri/1.6.6/tauri/struct.App.html#method.run_iteration)
should be able to handle this, but instead
[it busy-loops](https://github.com/tauri-apps/tauri/issues/8631).

Commenting on these issues, lead Tauri developer Fabian said he also expects
better from the underlying OSes and this causes him to
["cry (himself) to sleep at night 🤷"](https://github.com/tauri-apps/tauri/issues/8631#issuecomment-1898328992)
Me too, Fabian. Me too.

> ## Setting binary names is tricky

Reviewer: ```suggestion
### Setting binary names is tricky
```

On Linux our GUI is named `firezone-client-gui`, and so is our package.

We want:

1. The `.desktop` file to display our name as "Firezone Client", so it's
   human-friendly,
1. The package to be `firezone-client-gui`, so that it's distinct from our CLI
   Client
1. The GUI binary to be `/usr/bin/firezone-client-gui`,
1. And the service binary to be
   `/usr/lib/$ARCH/dev.firezone.client/firezone-client-ipc`, to keep it out of
   `$PATH`.

But Tauri's default is to name the first 3 the same and to put all binaries in
`/usr/bin`. We could fix this when we tear apart the `.deb` for the post-install
script, but I'd rather have a way to cut around the defaults and take control of
the bundler directly, like with the Wix files on Windows.

This is a case where Tauri doesn't reach "Easy things are simple, hard things
are possible."

> ## Initializing Tauri is flaky in CI... maybe?

Reviewer: ```suggestion
### Initializing Tauri is flaky in CI... maybe?
```

https://github.com/firezone/firezone/issues/3972#issuecomment-2010728424

> I think it's getting stuck on creating the GTK event loop for the window

At first we tried to initialize as much of the program as possible before
starting Tauri, and then move objects into the Tauri context. But we hit this
issue, which appears to be a race in initializing the GTK+ event loop or
something. It's arcane stuff 3 layers below us (Tauri, then `tao`, then GTK+) in
C code, so the workaround for now is to initialize Tauri immediately and do all
our own setup within the Tauri context.

It may not be Tauri's fault, but it's disappointing because it's nice to keep
control of `main` longer before handing control to Tauri and doing everything
inside callbacks.

> # Conclusion

Reviewer: ```suggestion
## Conclusion
```
